src/domain/auth/auth.repository.ts

```diff
@@ -1,26 +1,148 @@
+  /**
+   * Verifies the validity of a signed message and returns an access token:
+   *
+   * 1. Ensure the message itself has not expired.
+   * 2. Ensure the nonce was generated by us/is not a replay attack.
+   * 3. Verify the signature of the message.
+   * 4. Return an access token if all checks pass.
+   *
+   * @param args - DTO containing the message and signature to verify.
+   *
+   * The following adhere to JWT standard {@link https://datatracker.ietf.org/doc/html/rfc7519}
+   *
+   * @returns accessToken - JWT access token
+   * @returns tokenType - token type ('Bearer') to be used in the `Authorization` header
+   * @returns notBefore - epoch from when token is valid (if applicable, otherwise null)
+   * @returns expiresIn - time in seconds until the token expires (if applicable, otherwise null)
+   */
+  async verify(args: VerifyAuthMessageDto): Promise<{
+    accessToken: string;
+    tokenType: string;
+    notBefore: number | null;
+    expiresIn: number | null;
+  }> {
+    const isAuthorized = await this.isAuthorized(args).catch(() => false);
+
+    if (!isAuthorized) {
+      throw new UnauthorizedException();
+    }
+
+    const dateWhenTokenIsValid = args.message.notBefore
+      ? new Date(args.message.notBefore)
+      : null;
+    const dateWhenTokenExpires = args.message.expirationTime
+      ? new Date(args.message.expirationTime)
+      : null;
+
+    const secondsUntilTokenIsValid = dateWhenTokenIsValid
+      ? this.getSecondsUntil(dateWhenTokenIsValid)
+      : null;
+    const secondsUntilTokenExpires = dateWhenTokenExpires
+      ? this.getSecondsUntil(dateWhenTokenExpires)
+      : null;
+
+    const accessToken = this.jwtService.sign(args.message, {
+      ...(secondsUntilTokenIsValid !== null && {
+        notBefore: secondsUntilTokenIsValid,
+      }),
+      ...(secondsUntilTokenExpires !== null && {
+        expiresIn: secondsUntilTokenExpires,
+      }),
+    });
+
+    return {
+      tokenType: AuthService.AUTH_TOKEN_TOKEN_TYPE,
+      accessToken,
+      // Differing measurements match JWT standard {@link https://datatracker.ietf.org/doc/html/rfc7519}
+      notBefore: dateWhenTokenIsValid?.getTime() ?? null,
+      expiresIn: secondsUntilTokenExpires,
+    };
+  }
+
+  /**
+   * Verifies that a message is valid according to its expiration date,
+   * signature and nonce.
+   *
+   * @param args.message - SiWe message in object form
+   * @param args.signature - signature from signing the message
+   *
+   * @returns boolean - whether the message is valid
+   */
+  private async isAuthorized(args: {
     message: SiweMessage;
     signature: `0x${string}`;
   }): Promise<boolean> {
-    return this.authApi.verifyMessage(args);
+    const isExpired =
+      !!args.message.expirationTime &&
+      new Date(args.message.expirationTime) < new Date();
+
+    const isValidSignature = await this.authApi
+      .verifyMessage(args)
+      // Don't prevent nonce from being deleted
+      .catch(() => false);
```

Comment:
Nitpick: wdyt about using a `try-catch-finally` block here? The `nonce` deletion seems to fit in a `finally` clause at it needs to be always executed, and we can also have specific errors for each condition (expired, invalid nonce, invalid signature) that could be managed in a `catch` block.

Reply:
Refactored in 0ce020796f2e6da46916af649e7160d62810bf8e.